Modules/HTTP/Sources/HTTP/Parser/Parser.swift

```diff
@@ -1,56 +1,302 @@
+        process(state: .body, data: UnsafeBufferPointer<Int8>(start: data, count: length))
+        return 0
+    }
+    
+    private func processOnMessageComplete() -> Int32 {
+        process(state: .messageComplete)
+        return 0
+    }
+    
+    private func process(state newState: State, data: UnsafeBufferPointer<Int8>? = nil) {
+        if state != newState {
+            
+            switch state {
+            case .none, .messageBegin, .messageComplete:
+                break
+                
+            case .url:
+                let string = String(bytes: context.currentBuffer, encoding: String.Encoding.utf8)!
+                context.url = URL(string: string)!
+                
+            case .status:
+                let string = String(bytes: context.currentBuffer, encoding: String.Encoding.utf8)!
+                context.status = Response.Status(statusCode: Int(parser.status_code),
+                                                 reasonPhrase: string)
+                
+            case .headerField:
+                let string = String(bytes: context.currentBuffer, encoding: String.Encoding.utf8)!
+                context.currentHeaderField = CaseInsensitiveString(string)
+                
+                
+            case .headerValue:
+                let string = String(bytes: context.currentBuffer, encoding: String.Encoding.utf8)!
+                context.addValueForCurrentHeaderField(string)
+                
+            case .headersComplete:
+                context.currentHeaderField = nil
+                context.method = Request.Method(code: http_method(rawValue: parser.method))
+                context.version = Version(major: Int(parser.http_major), minor: Int(parser.http_minor))
+                
+            case .body:
+                context.body = context.currentBuffer
+                
+            }
+            
+            context.currentBuffer = DispatchData.empty
+            state = newState
+            
+            if state == .messageComplete {
+                let message: Message
+                switch mode {
+                case .request:
+                    var request = Request(method: context.method!,
+                                          url: context.url!,
+                                          headers: Headers(),
+                                          body: .buffer(context.body))
+                    request.headers = Headers(context.headers)
+                    
+                    message = request
+                case .response:
+                    let cookieHeaders =
+                        self.context.headers
+                            .filter { $0.key == "Set-Cookie" }
+                            .map { $0.value }
+                            .reduce(Set<String>()) { initial, value in
+                                return initial.union(Set(value.components(separatedBy: ", ")))
```

Comment:
As much as I love `filter`, `map`, `reduce` and friends @Danappelxx mentioned that they aren't as performant as simple `for in` loops 😞.